### DISCUSSION TITLE
BigQuery's Clustering/Partitioning Implementation

### DISCUSSION THREAD

**Discussion by pykenny** [sqlalchemy/sqlalchemy#8583] [Usage Questions] [answered]
I'm now trying to add [Clustering](https://cloud.google.com/bigquery/docs/creating-clustered-tables) and [Partitioning](https://cloud.google.com/bigquery/docs/creating-partitioned-tables) to [BigQuery's dialect](https://github.com/googleapis/python-bigquery-sqlalchemy). Both act as table optimization similar to indexing, whereas they're having constraints that are not compatible with `schema.Index` class's interface. The biggest one is that they can only be declared within table creation DDL (check out the SQL examples in the above documentation links for samples), and table altering on clustering/partitioning setup is not supported in their SQL syntax, hence overriding `schema.Index` class doesn't seem to be the way to go (it has `create()` method, which conflicts with the fact that it's impossible to create partitioning/clustering on existing table), and new class(es) might be needed instead, with similar usage as the original `Index` class.

I've checked out schema.Index and their parent classes, but still feel very confused about designing the class precisely (even haven't touched the details on query generation/traversal part).

There's a request about supporting these features on [BigQuery dialect's issue board](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/395) but haven't see anyone taking a stab on implementation, and I also list out several constraints in contrast to regular SQL indexing in [comment](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/395#issuecomment-1257774911).

So ideally their usage should look like this in imperative or declarative bases:
```python
table_01 = Table(
  "some_table", metadata,
  Column("id", Integer(), primary_key=True),
  Column("name", String()),
  Column("created_at", DateTime()),
  bigquery_clustering=Clustering("id"),
  bigquery_partitioning=Partitioning(
    date_trunc("created_at", TimeScale.DAY),
    required_in_filter=True,
    partition_expire_days=3,
  ),
)

# or

table_02 = Table(
  "some_table", metadata
  Column("id", Integer(), primary_key=True),
  Column("name", String()),
  Column("created_at", DateTime()),
)
clustering_setup = Clustering("id", table=table_02)
partitioning_setup = Partitioning(
  date_trunc("created_at", TimeScale.DAY),
  table=table_02,
  required_in_filter=True,
  partition_expire_days=3,
)

# or

class SomeTable(ModelBase):
  __tablename__ = "some_table"
  __table_args__ = {
    "bigquery_clustering": Clustering("id"),
    "bigquery_partitioning": Partitioning(
      date_trunc("created_at", TimeScale.DAY),
      required_in_filter=True,
      partition_expire_days=3,
    ),
  }
  id_=Column("id", Integer(), primary_key=True)
  name=Column(String())
  created_at=Column(DateTime())
```

Expected schema output:
```sql
CREATE TABLE mydataset.some_table
 id INT64,
 name STRING
CLUSTER BY id
PARTITION BY DATETIME_TRUNC(created_at, DAY)
OPTIONS (
  partition_expiration_days = 3,
  require_partition_filter = TRUE
)
```

Seems that `SchemaItem` is required for every schema defining components, and `ColumnCollectionMixin` may be needed to match column/names passed to argument list, but I'm not not sure if that suffices to meet the requirements.

**Comment by zzzeek** [comment, marked as answer]
hey there-

PARTITION BY , this is a big deal, as it impacts much beyond the schema level DDL itself; ~~these structures can be referenced in queries which is part of what makes them useful, not to mention their syntax expands out to a few places such as within primary key constraints.~~ (Edit, Feb 2025, Mike seems to be thinking of temporal tables / queries here)

~~as such, we are adding this feature natively~~. (edit: Mike in Feb 2025 does not understand what MIke in 2022 is talking about here, he seems to have had some idea about temporal table support and partitioning support being under some common API...but you can add keywords like "PARTITION BY" to a CREATE TABLE statement pretty easily right now anyway, it doesnt need some big new API...mysql dialect does these already).

 However, it is complicated and we were at the point where we were considering pushing it to release 2.1 which would be a year out, however, this was based on the fact that only one user has expressed any interest in this feature.    The work in progress for it, which I plan to modify at some point for inclusion, can be seen at:

https://github.com/sqlalchemy/sqlalchemy/issues/4797
https://github.com/sqlalchemy/sqlalchemy/pull/7933
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803

my alternative plan, once 2.0 beta1 is released and we see evidence (of which this discussion is evidence) that we need to push for the PARTITION BY API sooner, is that I will work on it during the 2.0 beta series, or if I can get it finished up before beta1 in the next few weeks.

So any work BQ does with PARTITION BY would want to be tuned into this.    That said, if BQ is looking for this feature implemented in a more expedient fashion and especially in the 1.4 series, if you need only the schema-level DDL constructs and nothing else I'd likely stick with the dialect kwargs version, because that will keep your implementation distinct from what we eventually add here, that is, "bigquery_partitioning" keyword argument.      It's of course the best outcome if we all just focus on the actual feature at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3803, make sure what we're doing there covers BQ's cases, and then we all get 2.0 with "PARTITION" support out the door.

re: clustering, I'm less familiar with this concept as expressed in BQ, although keeping it within "bigquery_clustering" will keep you the most isolated from other SQLAlchemy dialects and future APIs.    Also, the general use of the dialect kwargs has the most straightforward developer API; having real constructs that get added to the `Table(*args)` is much more complicated from an implementation point of view and generally it's likely best if this approach is not used by third party dialects in most (if not all) cases.

Within the scope of using "bigquery_partitioning" and "bigquery_clustering" kw arguments you might want to consider a more lightweight API that doesn't require the use imported classes like `Partitioning` and `Clustering`, unless these classes have additional uses within querying.

**Reply by pykenny** [reply]
@zzzeek Unlike in other services, which usually refers to data distributing mechanisms, clustering setup in BigQuery pre-sorts records within a table (or within each partition when partitioning is enabled) by value of specified column(s), hence filtering/joining/sorting on clustered columns becomes more efficient ideally.

**Reply by zzzeek** [reply]
I'm sure it's a great feature!

**Reply by pykenny** [reply]
> having real constructs that get added to the `Table(*args)` is much more complicated from an implementation point of view

Since I'm only using declarative models to describe tables, my final goal is to make the third example for defining `some_table` available in my codebase. Also updated `__table_args__` usage in that example to make it correctly reflect the first example (imperative approach within `Table` construct).

**Reply by pykenny** [reply]
Clustering only requires assigning target columns, so it's possible to make it without introducing constructor (such as passing string collection). Partitioning has two additional options but breaking one kw argument into three (partitioning criteria, partition expiration period in days, enabling filtering requirement when used in query) looks alright to me.

**Reply by zzzeek** [reply]
do you have interest in targeting the actual partitioning API I plan to have in the 2.x series ?

**Reply by pykenny** [reply]
Sure! Would be happy to take a look and to see if it's possible to make 2.x's partitioning syntax compatible with BQ's.

As I'm using 1.4 in my codebase, what's the best approach to work on both sides at the same time?

**Reply by zzzeek** [reply]
well for now just do the thing you're going to do that works in 1.4.    I think mostly, if you were to look at the new work and see that it covers your cases also or if we are missing something huge

**Reply by pykenny** [reply]
Got it. Just found that [the library fills all additional table settings through `post_table_creation()`](https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/sqlalchemy_bigquery/base.py#L632-L652) so I'll start from overriding the method (insert clustering/partitioning settings before `OPTIONS` part).

**Reply by pykenny** [reply]
@zzzeek Another question popped up when looking into BQ's partitioning setup:

In partitioning, BigQuery allows generating partitions based on ingestion time auto-generated within their service. When enabled, BigQuery creates a hidden column for storing these timestamps (already truncated to the precision specified for partitioning) as reference. This column is absent from column list of the table's schema, but still accessible through `_PARTITIONTIME` (`TIMESTAMP`) or `_PARTITIONDATE` (`DATE`) names within column selection and filters . Not really certain about abstracting such feature in a table/model.

Example:
```sql
CREATE TABLE
  mydataset.newtable (transaction_id INT64)
PARTITION BY
  _PARTITIONDATE
```

```sql
SELECT
  id,
  DATE_TRUNC(_PARTITIONDATE, YEAR),
  TIMESTAMP_TRUNC(_PARTITIONTIME, WEEK)
FROM
  mydataset.newtable
WHERE
  _PARTITIONDATE >= '2022-01-01'
```

**Reply by zzzeek** [reply]
right so this is also to be part of our new feature.   for now, you can get a similar effect by adding a Column to a Table with the [system=True](https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=column+system#sqlalchemy.schema.Column.params.system) parameter.  This will mark the column as implicitly present, without it being part of any CREATE TABLE statement.

**Reply by pykenny** [reply]
We finally ended up with (non-trivial) solution using dialect options, and plug it into the original compiler/dialect. Just as discussed above, these setups are processed and appended to the table creation statement through `post_table_creation()` when provided in table arguments.

**Reply by CaselIT** [reply]
is the code public? if so a link to it could be useful to see how this feature is used when designing the general api in sqlalchemy

**Reply by pykenny** [reply]
@CaselIT It's now developed under a private repo, but we're considering make it a PR at BigQuery's dialect plugin to receive more feedback.

**Reply by hrshdhgd** [reply]
Hello! Has there been any progress on this feature being available in `sqlalchemy`?  I have a use-case that'll benefit tremendously if this were available.

**Reply by zzzeek** [reply]
hi unfortunately adding temporal table support was abandoned due to the ratio of complexity (massively, extremely high) to user interest (vanishing towards zero), so I would seek to build third party extensions for now

however there is more extensibility coming in SQLAlchemy 2.1 in #12195 which might make it easier to build such extensions

**Reply by pykenny** [reply]
As for BigQuery dialect library, table partitions/clustered order support has been added to later version - we’re still sticking to our internal patch implementation, though

 - https://github.com/googleapis/python-bigquery-sqlalchemy/pull/928

**Reply by zzzeek** [reply]
yah im not sure what im thinking in my eaerlier comments here, temporal tables and partitioning seem to be different concerns so im not sure if I thought these could be unified in some way, but in any case, we already support emitting keywords like "PARTITION BY" for MySQL and others right now, this is not a big deal by itself

**Reply by hrshdhgd** [reply]
> we already support emitting keywords like "PARTITION BY" for MySQL and others right now, this is not a big deal by itself

Oh interesting, could you please direct me to the documentation for this? I was looking and have been unsuccessful.

**Reply by hrshdhgd** [reply]
All I could find was https://docs.sqlalchemy.org/en/20/changelog/migration_13.html#dialect-improvements-and-changes-postgresql which seems PostGRES specific.

**Reply by CaselIT** [reply]
interesting. seems not. It was added here: 93b11905e599a6d73a85d2085e15385ebf46cdc6

**Reply by hrshdhgd** [reply]
Thanks @CaselIT ! That gives me some hope. Few follow up questions:
 - The PR only addresses `PARTITION BY`. Has there been development for `CLUSTER BY` too? (Hopefully in another PR 😄 )

**Reply by CaselIT** [reply]
i'm not familiar with that syntax, but if it's a create table option you should be able to specify it already using `mysql_...=...`

**Reply by hrshdhgd** [reply]
Makes sense. Also the use case I have is for use in BigQuery. The syntax would not be specific to MySQL, correct? Just want to make sure.

**Reply by CaselIT** [reply]
for bigquery would would need to use `bigquery_...` as keyword option in the table

**Reply by hrshdhgd** [reply]
So I wrote a small test that passes:
```python
import pytest

from sqlalchemy import create_engine, Table, Column, Integer, String, Date, MetaData
from sqlalchemy.schema import CreateTable
from sqlalchemy_bigquery import BigQueryDialect  # Native BigQuery support

def test_bigquery_partition_cluster_options():
    metadata = MetaData()

    # Define table with partitioning and clustering
    table = Table(
        "test_table",
        metadata,
        Column("id", Integer, primary_key=True),
        Column("name", String),
        Column("created_at", Date),
        bigquery_partitioning="DATE(created_at)",
        bigquery_cluster_by=["name", "id"],
    )

    # Verify partitioning and clustering options exist
    assert "bigquery" in table.dialect_options
    assert table.dialect_options["bigquery"]["partition_by"] == "DATE(created_at)"
    assert table.dialect_options["bigquery"]["cluster_by"] == ["name", "id"]

    ddl = str(CreateTable(table).compile(dialect=BigQueryDialect()))
    print(ddl)

if __name__ == "__main__":
    pytest.main([__file__])
```
But the print statement gives me just this:
    
```sql
    CREATE TABLE `test_table` (
        `id` INT64 NOT NULL, 
        `name` STRING, 
        `created_at` DATE
)
```
neither `PARTITION BY` nor `CLUSTER BY` get translated into the DDL.
with the following versions:
```shell
sqlalchemy                2.0.38                   pypi_0    pypi
sqlalchemy-bigquery       1.12.1                   pypi_0    pypi
```

Am I doing something wrong?

**Reply by CaselIT** [reply]
have you verified that they are supported by bigquery? I believe that they need to be explicitely supported by the dialect to be used in the ddl compilation

**Reply by hrshdhgd** [reply]
Aah, that is a good point. I will investigate that. Thank you for your patience!

**Reply by hrshdhgd** [reply]
I had the attributes wrong. Just so that I end the loop this works:

```python
from sqlalchemy import create_engine, Table, Column, Integer, String, Date, MetaData
from sqlalchemy.schema import CreateTable
from sqlalchemy_bigquery import BigQueryDialect  # Native BigQuery support
from google.cloud import bigquery

def test_bigquery_partition_cluster_options():
    metadata = MetaData()

    # Define table with partitioning and clustering
    table = Table(
        "test_table",
        metadata,
        Column("id", Integer, primary_key=True),
        Column("name", String),
        Column("created_at", Date),
        Column("zipcode", Integer),
        bigquery_time_partitioning=bigquery.TimePartitioning(
            field="created_at",
            type_="MONTH",
            expiration_ms=1000 * 60 * 60 * 24 * 30 * 6, # 6 months
        ),


        bigquery_require_partition_filter=True,
        
        
        bigquery_clustering_fields=["name", "id"],
    )

    ddl = str(CreateTable(table).compile(dialect=BigQueryDialect()))
    print(ddl)

if __name__ == "__main__":
    pytest.main([__file__])
```

produces

```sql
CREATE TABLE `test_table` (
        `id` INT64 NOT NULL, 
        `name` STRING, 
        `created_at` DATE, 
        `zipcode` INT64
) PARTITION BY DATE_TRUNC(created_at, MONTH)
CLUSTER BY name, id
OPTIONS(partition_expiration_days=180.0, require_partition_filter=true)
```

Thank you @CaselIT !!
